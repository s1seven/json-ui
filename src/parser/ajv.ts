import Ajv2019 from "ajv/dist/2019";
import draft7MetaSchema from "ajv/dist/refs/json-schema-draft-07.json";
import { AJV_ALLOWED_KEYWORDS } from "../constants";
import addFormats from "ajv-formats";

export const ajv = new Ajv2019();

export const ajvFactory = () => {
	const ajv = new Ajv2019();
	ajv.addMetaSchema(draft7MetaSchema);
	addFormats(ajv);
	AJV_ALLOWED_KEYWORDS.forEach((keyword) => {
		ajv.addKeyword(keyword);
	});
	return ajv;
};

addFormats(ajv);
ajv.addMetaSchema(draft7MetaSchema);

AJV_ALLOWED_KEYWORDS.forEach((keyword) => {
	ajv.addKeyword(keyword);
});
