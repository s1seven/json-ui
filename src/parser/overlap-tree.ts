import { JSONSchema7 } from "json-schema";
import { inferType } from "./infer";
import { ajvFactory } from "./ajv";
import { get } from "lodash";

export type TreeItem = [
  label: string,
  isValidOverlap: boolean,
  children: TreeItem[]
];

const ajv = ajvFactory();

export const overlapTree = (
  schemas: (JSONSchema7 | undefined)[],
  key = ""
): TreeItem => {
  if (schemas.includes(void 0)) return [key, false, []];
  if (schemas.length < 1) throw new Error("No schemas provided.");
  const schema = schemas[0]!;
  const type = inferType(schema);

  const mergedSchema = {
    allOf: schemas,
  };

  const isValid = ajv.validateSchema(mergedSchema) as boolean;
  console.log(isValid, JSON.stringify(mergedSchema));

  return [
    key,
    isValid,
    type === "object"
      ? Object.keys(schema.properties!).map((key) =>
          overlapTree(
            schemas.map((scm) =>
              get(scm, `properties.${key}`)
            ) as JSONSchema7[],
            key
          )
        )
      : [],
  ];
};
