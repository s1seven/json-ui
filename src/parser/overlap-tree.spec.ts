import { JSONSchema7 } from "json-schema";
import { overlapTree } from "./overlap-tree";
import { ajvFactory } from "./ajv";

test("simple tree overlap", async () => {
  const ajv = ajvFactory();

  console.log(
    "hello",
    ajv.validateSchema({
      $schema: "http://json-schema.org/draft-07/schema#",
      type: "objesdct",
      properties: {
        example: {
          type: "string",
          minLength: 10,
          maxLength: 5,
        },
      },
      required: ["missingProperty"],
    }),
    ajv.errors
  );

  const schemas: JSONSchema7[] = [
    {
      type: "object",
      properties: {
        brand: {
          type: "string",
        },
        model: {
          type: "string",
        },
      },
    },
    {
      type: "object",
      properties: {
        brand: {
          type: "object",
        },
        model: {
          type: "string",
        },
      },
    },
  ];
  const tree = overlapTree(schemas);
  expect(tree).toMatchSnapshot();
});

// export const asdasd: TreeItem = [
//   "Root",
//   true,
//   [
//     ["Root.1", true],
//     ["Root.2", true, [["Root.2.1", true]]],
//     ["Root.1", true],
//   ],
// ];
