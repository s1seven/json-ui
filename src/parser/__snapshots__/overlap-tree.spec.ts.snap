// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple tree overlap 1`] = `
[
  "",
  true,
  [
    [
      "brand",
      true,
      [],
    ],
    [
      "model",
      true,
      [],
    ],
  ],
]
`;
